src/Ouroboros_Praos_Implementation.thy

```diff
@@ -5,7 +5,7 @@ theory Ouroboros_Praos_Implementation
     Finite_Map_Extras
```

Comment:
It seems you’re using a local version of `Finite_Maps_Extras` here. Could you please change the import and the `ROOT` file such that `Finite_Map_Extras` is taken from the AFP? Once you’ve made the change, I can continue to review this pull request.

Reply:
In fact, the Praos specification is currently not mentioned in the `ROOT` file (I once made the `ROOT` file a stub, so that non-buildable theories wouldn’t cause automated builds to fail). Could you perhaps add the Praos specification to the `ROOT` file?

Reply:
Could you place the import of `Finite_Map_Extras` after the `HOL-Library` imports? This is more sensible, in my opinion, as `Finite_Map_Extras` itself depends on bits from `HOL-Library` and is thus on a higher level, so to say.

Reply:
Since you import stuff from `HOL-Library`, you should also mention `HOL-Library` in the `ROOTS` file, right?

Reply:
I also think the mention of `Finite_Map_Extras` in the `ROOTS` file should come before the mention of `Thorn_Calculus`, for analogous reasons as above.

Reply:
> It seems you’re using a local version of Finite_Maps_Extras here. Could you please change the import and the ROOT file such that Finite_Map_Extras is taken from the AFP? Once you’ve made the change, I can continue to review this pull request.

Fixed in b98dd83aa4ffc241b820080467c695d050e5e988, 7b112ac5ef721b30a068d5c340a5ef7935cd38fb, 7f5306a1062c8e7e66f0165641fd6a65c3dde296, c47f6d3f864f61d07822dde91439e411c4ce8e86, b6d047e1693eff4db71b08a9de34bd214856e0d6, and 50a86cd838865abe00631f0f4bcdcc1b6d7c0232. Note that there is an issue when building `Finite-Map-Extras`, which we are currently trying to work around, as explained further on.

Reply:
>Could you place the import of Finite_Map_Extras after the HOL-Library imports? This is more sensible, in my opinion, as Finite_Map_Extras itself depends on bits from HOL-Library and is thus on a higher level, so to say.

Fixed in adc84c8f2f89f44de4ca94c3c2201bf405448884, please check.

Reply:
> Since you import stuff from `HOL-Library`, you should also mention `HOL-Library` in the `ROOTS` file, right?

By "`ROOTS` file" do you actually mean the `ROOT` file?

Reply:
Yes and yes.

Reply:
> I also think the mention of `Finite_Map_Extras` in the `ROOTS` file should come before the mention of `Thorn_Calculus`, for analogous reasons as above.

Fixed in 087275a4b2bba451952c5bd049f7e38286110812, please check.

Reply:
> Since you import stuff from `HOL-Library`, you should also mention `HOL-Library` in the `ROOTS` file, right?

Fixed in 83325358ec1a9ebc97b0b417e397aeee7bef313b, please check.